examples/autoignition-mpi.py

```diff
@@ -223,25 +222,57 @@ def my_rhs(t, state):
                 + eos.get_species_source_terms(state))
 
     def my_checkpoint(step, t, dt, state):
-        reaction_rates = eos.get_production_rates(state)
-        viz_fields = [("reaction_rates", reaction_rates)]
-        return sim_checkpoint(discr, visualizer, eos, cv=state,
-                              vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
+        from mirgecom.simutil import check_step
+        do_status = check_step(step=step, interval=nstatus)
+        do_viz = check_step(step=step, interval=nviz)
+        do_health = check_step(step=step, interval=nhealth)
+
+        if do_status or do_viz or do_health:
+            dv = eos.dependent_vars(state)
+            reaction_rates = eos.get_production_rates(state)
+            io_fields = [
+                ("cv", state),
+                ("dv", dv),
+                ("reaction_rates", reaction_rates)
+            ]
+
+        if do_status:  # This is bad, logging already completely replaces this
+            from mirgecom.io import make_status_message
+            status_msg = make_status_message(discr=discr, t=t, step=step, dt=dt,
+                                             cfl=current_cfl, dependent_vars=dv)
+            if rank == 0:
+                logger.info(status_msg)
+
+        errored = False
+        if do_health:
+            from mirgecom.simutil import check_naninf_local, check_range_local
+            if check_naninf_local(discr, "vol", dv.pressure) \
+               or check_range_local(discr, "vol", dv.pressure):
+                errored = True
+                message = "Invalid pressure data found.\n"
+            comm = discr.mpi_communicator
+            if comm is not None:
+                errored = comm.allreduce(errored, op=MPI.LOR)
+            if errored:
+                if rank == 0:
+                    logger.info("Fluid solution failed health check.")
+                logger.info(message)   # do this on all ranks
```

Comment:
#384

Reply:
> I don't think this works in its current state. You'll need to log the message from one of the ranks that produced an error. `message` isn't defined on all ranks.
> 

I cleaned this up a bit in (70ff6dc), and plan on using #384. But I do think that using something like `logger.info(health_message)` (or similar) on every rank will be better for capturing all the messages from every rank.  These will only come out if something goes awry, and it is often useful to have the errors from all the ranks (especially since we don't currently have a mechanism to tell us where solution errors occur like we had in PC2).

Reply:
I would probably use the exception for the summary and add logging for details. I think the logging can be done separately from the synchronization stuff since it's coming out on each rank independently anyway. Something like:

```python
if check_naninf_local(discr, "vol", dv.pressure):
    logger.info(f"Detected NaN on rank blah, location blah blah.")
    exception = SimulationHealthError("NaNs detected in pressure.")
```
?

Reply:
> I would probably use the exception for the summary and add logging for details. I think the logging can be done separately from the synchronization stuff since it's coming out on each rank independently anyway. Something like:
> 
> ```python
> if check_naninf_local(discr, "vol", dv.pressure):
>     logger.info(f"Detected NaN on rank blah, location blah blah.")
>     exception = SimulationHealthError("NaNs detected in pressure.")
> ```
> 
> ?

I like how you've eliminated the dumb accumulated message a lot. 

From a user's perspective, they are doing more checks down the line that they also want to know the results of, so could we just key on an integer `error` variable about whether *any* error occurred and then throw the exception at the end with a more general message?  ("Failed health checking"), or some such?  Then if any rank had the health exception, they would all terminate, and the details of the error would be in the logs. (One thing to note: If we also had a `check_naninf` that just did the `grudge.nodal_max` on the back-end, then the pressure check here would already be global.  I argue that it is not our business to prescribe which of these the user should use in their own health check.)

An additional question/detail about this: If we commit to allowing rank-specific health-checking (or error handling in general) as has been suggested or at least implied, then can't we synchronize across ranks easily with a single allreduce of a single integer, instead of one that exchanges the whole exception?  Isn't this just another detail that should be left to the user's discretion?   

The pre and post-step callbacks are a nice place to do this type of error checking because they are already collective, and offers the user a major advantage of the ability to define (through code that they write and maintain) what happens when those errors occur.

Reply:
> I like how you've eliminated the dumb accumulated message a lot.
> 
> From a user's perspective, they are doing more checks down the line that they also want to know the results of, so could we just key on an integer `error` variable about whether _any_ error occurred and then throw the exception at the end with a more general message? ("Failed health checking"), or some such? Then if any rank had the health exception, they would all terminate, and the details of the error would be in the logs. (One thing to note: If we also had a `check_naninf` that just did the `grudge.nodal_max` on the back-end, then the pressure check here would already be global. I argue that it is not our business to prescribe which of these the user should use in their own health check.)

:man_shrugging: I like having the exceptions give some indication of what went wrong. From there you can either scroll up to get the details or go straight to visualizing.

> An additional question/detail about this: If we commit to allowing rank-specific health-checking (or error handling in general) as has been suggested or at least implied, then can't we synchronize across ranks easily with a single allreduce of a single integer, instead of one that exchanges the whole exception? Isn't this just another detail that should be left to the user's discretion?

Answered [here](https://github.com/illinois-ceesd/mirgecom/pull/384#issuecomment-860889833).